Make sure to use override where appropriate

masterleinad opened the pull request:
This fixes a pet peeve of mine. Namely, this PR marks all `virtual` functions in derived classes as `override` and adds the flag `Wsuggest-override` for `gcc`.
This builds for me using `gcc-8`with all dependencies (apart from CUDA) enabled and forcing bundled packages. Only the bundled `boost`needed to be fixed as well, while using `boost-1.67` was just fine.
I didn't add any `virtual`specifiers but made sure to not remove any as well.

drwells commented:
Thanks for fixing this.

Please remove `examples/step-40.ninja`. Otherwise I think we should merge.

masterleinad commented:
> Please remove examples/step-40.ninja. Otherwise I think we should merge.

Done

masterleinad commented:
/run-tests

bangerth commented:
How on earth did you write this patch without going crazy?

masterleinad commented:
> How on earth did you write this patch without going crazy?

I modified the `clang-tidy` `modernize-use-override` check implementation to not delete the `virtual` specifier and fixed the rest manually.

bangerth commented on include/deal.II/numerics/dof_output_operator.templates.h:
```diff
@@ -66,7 +66,7 @@ namespace Algorithms
     out.clear ();
     return *this;
-  }
+  };
```
This semicolon now produces a warning for me. Would you mind removing it again?
